### PR TITLE
[codex] Fix sentence-transformers 5 compatibility

### PR DESCRIPTION
**Pull request by ranieri**
## Summary
- update the custom asm sentence-transformers wrapper to follow the sentence-transformers 5.x input module contract
- replace the broken pretokenized tokenizer path with ids-first encoding plus tokenizer-managed padding
- restore tied embedding weights from checkpoint load so ARM64BERT embeddings remain stable across reloads

## Root Cause
sentence-transformers 5.x changed the Transformer/tokenizer contract, and the upstream transformers loader now leaves the tied word embedding parameter uninitialized for this checkpoint layout.

## Validation
- `pdm run test`

### CODE REVIEW

File: citatio/pyproject.toml
Comment (akaIDIOT) [inline citatio/pyproject.toml:1]: We can lift this change to main, right? This need not be part of sentence-transformers 4 → 5 🤔

**Comment by ranieri** [comment]
I've tested `pretrain.py` is functional. 
I know `finetune.py` is broken.

File: asmtransformers/asmtransformers/models/asmbert.py [hunk 1]
```diff
@@ -1,7 +1,9 @@
 import json
+import warnings
 from copy import deepcopy
 
 import torch
+from torch import nn
 from torch.nn import CrossEntropyLoss
```
Comment (akaIDIOT) [inline asmtransformers/asmtransformers/models/asmbert.py]: These two can likely be combined, I'd expect.

File: asmtransformers/asmtransformers/models/asmbert.py [hunk 2]
```diff
@@ -20,15 +22,59 @@ class ASMBertModel(BertModel):
...
     """
 
-    _tied_weights_keys = [*(BertModel._tied_weights_keys or []), 'embeddings.position_embeddings.weight']
+    _tied_weights_keys = {
+        **(BertModel._tied_weights_keys or {}),
+        'embeddings.position_embeddings.weight': 'embeddings.word_embeddings.weight',
+    }
 
-    def __init__(self, config, add_pooling_layer=True):
+    def __init__(self, config, add_pooling_layer=True, *, delay_tie_for_load=False):
```
Comment (akaIDIOT) [inline asmtransformers/asmtransformers/models/asmbert.py:28]: KW-only is great, should we include `add_pooling_layer` with this, or is it required to keep that around as a positional for something else?

File: asmtransformers/asmtransformers/models/asmbert.py [hunk 3]
```diff
@@ -20,15 +22,59 @@ class ASMBertModel(BertModel):
...
+                UserWarning,
+                stacklevel=2,
+            )
+            model.embeddings.word_embeddings.weight.data.copy_(model.embeddings.position_embeddings.weight.data)
+            loading_info['missing_keys'].discard('embeddings.word_embeddings.weight')
+        model.tie_shared_embeddings()
+
+        if output_loading_info:
+            return model, loading_info
+        return model
```
Comment (akaIDIOT) [inline asmtransformers/asmtransformers/models/asmbert.py:73]: For the record: I hate this construction, but it seems part of `super()`'s contract, so ... fine 😤

File: asmtransformers/asmtransformers/models/asmbert.py [hunk 4]
```diff
@@ -198,26 +247,26 @@ def set_tokenizer(self, tokenizer):
...
-            truncation=True,
-            padding=True,
+            encoded_inputs.append(
+                {
+                    # The assembly preprocessor already splits the function into model vocabulary tokens.
+                    'input_ids': self.tokenizer.convert_tokens_to_ids(tokens[:512]),
+                }
+            )
+
+        return self.tokenizer.pad(
```
Comment (akaIDIOT) [inline asmtransformers/asmtransformers/models/asmbert.py:261]: As this is calling `pad()`, is the `+= self.padding[...]` above needed? What's the difference?

File: asmtransformers/asmtransformers/models/asmsentencebert.py [hunk 5]
```diff
@@ -26,30 +27,58 @@ def __init__(
...
         nn.Module.__init__(self)
 
-        if model_args is None:
-            model_args = {}
-
-        model = ASMBertModel.from_pretrained(model_name_or_path, **model_args)
-        model.tokenizer = tokenizer
-        self.tokenizer = tokenizer
-
+        model = ASMBertModel.from_pretrained(model_name_or_path, **(model_args or {}))
```
Comment (akaIDIOT) [inline asmtransformers/asmtransformers/models/asmsentencebert.py:30]: This is better than the `if is None` we had 👌🏻

File: asmtransformers/asmtransformers/models/asmsentencebert.py [hunk 6]
```diff
@@ -26,30 +27,58 @@ def __init__(
...
+
+    @property
+    def tokenizer(self):
+        return self.processor
+
+    @tokenizer.setter
+    def tokenizer(self, tokenizer):
+        self.processor = tokenizer
+        if hasattr(self, 'model'):
+            self.model.tokenizer = tokenizer
```
Comment (akaIDIOT) [inline asmtransformers/asmtransformers/models/asmsentencebert.py:81]: Considering ST5 has renamed these things: what code still needs there to be a `tokenizer` attribute, our own or someone else's? (in the second case: could we just ditch the property?)

File: asmtransformers/asmtransformers/models/asmbert.py [hunk 7]
```diff
@@ -198,26 +247,26 @@ def set_tokenizer(self, tokenizer):
...
-
-        tokenized = self.tokenizer(
-            tokens_batch,
-            # input is pre-tokenized by the preprocessor
-            is_split_into_words=True,
-            add_special_tokens=False,
-            # truncate the encoded sequence to a maximum of max_length tokens
-            truncation=True,
-            padding=True,
+            encoded_inputs.append(
```
Comment (akaIDIOT) [inline asmtransformers/asmtransformers/models/asmbert.py:254]: Not a fan of the init + append style here, but the old code did that too. Let's fix that later.

**Review by akaIDIOT — APPROVED** [review]

File: asmtransformers/asmtransformers/models/asmbert.py (hunk 2 above)
Comment (ranieri) [reply asmtransformers/asmtransformers/models/asmbert.py:28]: That's part of the `BertModel.__init__` contract.

File: asmtransformers/asmtransformers/models/asmsentencebert.py (hunk 6 above)
Comment (ranieri) [reply asmtransformers/asmtransformers/models/asmsentencebert.py:81]: A lot of the logic surrounding the tokenizer can be simplified substantially, but I'm leaving this for a future PR

File: asmtransformers/asmtransformers/models/asmbert.py (hunk 4 above)
Comment (ranieri) [reply asmtransformers/asmtransformers/models/asmbert.py:261]: Good catch! The padding tokes we were adding manually are also interfering with the attention mask.

Comment (ranieri) [reply asmtransformers/asmtransformers/models/asmbert.py:261]: Further inspection reveals this was present in main as well. I'm making an issue and leaving it for a future PR

File: asmtransformers/asmtransformers/models/asmbert.py (hunk 1 above)
Comment (drxzcl) [reply asmtransformers/asmtransformers/models/asmbert.py]: Done!
